solidity/contracts/bridge/library/FraudVerification.sol

```diff
@@ -0,0 +1,285 @@
+        uint256 fraudChallengeDepositAmount; //TODO: Initialize
+    }
+
+    struct FraudChallenge {
+        address challenger;
+        uint256 ethDepositAmount;
+        uint32 reportedAt;
+        bool defended;
+    }
+
+    struct Data {
+        Parameters parameters;
+        /// Collection of all submitted fraud challenges indexed by challenge
+        /// key built as keccak256(walletPublicKey|sighash|v|r|s).
+        mapping(uint256 => FraudChallenge) fraudChallenges;
+    }
+
+    event FraudChallengeDepositAmountUpdated(
+        uint256 newFraudChallengeDepositAmount
+    );
+
+    event FraudChallengeSubmitted(
+        bytes walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    );
+
+    event FraudChallengeDefended(
+        bytes walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    );
+
+    /// @notice Submits a wallet fraud challenge.
+    /// @param walletPublicKey The public key of the wallet in the uncompressed
+    ///                        and unprefixed format (64 bytes).
+    /// @param sighash Hash of the data (preimage) used for generating signature
+    ///                for the given input.
+    /// @param v Signature recovery value.
+    /// @param r Signature r value.
+    /// @param s Signature s value.
+    function submitFraudChallenge(
+        Data storage data,
+        bytes memory walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        // TODO: Bitcoin uses signatures in the DER format. From a signature in
+        // DER form values r and s can be easily extracted. Check how to obtain
+        // the v value. It can take values: 27, 28, 29, 30, 31, 32, 33, 34.
+        // https://bitcoin.stackexchange.com/questions/38351/ecdsa-v-r-s-what-is-v
+        require(
+            msg.value >= data.parameters.fraudChallengeDepositAmount,
+            "The amount of ETH deposited is too low"
+        );
+
+        bool verificationResult =
+            CheckBitcoinSigs.checkSig(walletPublicKey, sighash, v, r, s);
+        require(verificationResult, "Signature verification failure");
+
+        uint256 challengeKey =
+            uint256(
+                keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+            );
+
+        FraudChallenge storage fraudChallenge =
+            data.fraudChallenges[challengeKey];
+        require(fraudChallenge.reportedAt == 0, "Fraud already challenged");
+
+        fraudChallenge.challenger = msg.sender;
+        fraudChallenge.ethDepositAmount = msg.value;
+        /* solhint-disable-next-line not-rely-on-time */
+        fraudChallenge.reportedAt = uint32(block.timestamp);
+        fraudChallenge.defended = false;
+
+        // TODO: Consider emitting the event with walletPublicKey in the
+        //       compressed format as it's how we identify wallets in the Bridge.
```

Comment:
I think we can use the `walletPublicKeyHash` as we need to compute/pass it anyway. This is `bytes20` so should be cheaper to use as event parameter than `walletPublicKey` which is `bytes`

Reply:
All functions calculate and  emit the `walletPublicKeyHash` now. Done in 3139da05e1fa4f1c5f8482a72f93fc35d87ac4b2.